File: packs/dsa-linked-lists-py/challenges/05-remove-linked-list-elements.test.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def to_list(head):
    result = []
    while head:
        result.append(head.val)
        head = head.next
    return result


def from_list(lst):
    dummy = ListNode(0)
    curr = dummy
    for val in lst:
        curr.next = ListNode(val)
        curr = curr.next
    return dummy.next


from solution import remove_elements


def test_remove_middle_and_end():
    head = from_list([1, 2, 6, 3, 4, 5, 6])
    result = remove_elements(head, 6)
    assert to_list(result) == [1, 2, 3, 4, 5]


def test_remove_from_empty():
    result = remove_elements(None, 1)
    assert result is None


def test_remove_all():
    head = from_list([7, 7, 7, 7])
    result = remove_elements(head, 7)
    assert result is None


def test_remove_head():
    head = from_list([1, 1, 2, 3])
    result = remove_elements(head, 1)
    assert to_list(result) == [2, 3]


def test_remove_none_present():
    head = from_list([1, 2, 3])
    result = remove_elements(head, 4)
    assert to_list(result) == [1, 2, 3]

File: packs/dsa-linked-lists-ts/challenges/05-remove-linked-list-elements.test.ts
import { join } from 'node:path';
import { writeChallengeToTmp, importModule } from '../../test-helpers.js';

class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val = 0, next: ListNode | null = null) {
    this.val = val;
    this.next = next;
  }
}

function toList(head: ListNode | null): number[] {
  const result: number[] = [];
  while (head) {
    result.push(head.val);
    head = head.next;
  }
  return result;
}

function fromList(arr: number[]): ListNode | null {
  const dummy = new ListNode(0);
  let curr = dummy;
  for (const val of arr) {
    curr.next = new ListNode(val);
    curr = curr.next;
  }
  return dummy.next;
}

const CHALLENGE = join(import.meta.dirname, '05-remove-linked-list-elements.json');

describe('Remove Linked List Elements', () => {
  let removeElements: (head: ListNode | null, val: number) => ListNode | null;
  let cleanup: () => void;

  beforeAll(async () => {
    const tmp = writeChallengeToTmp(CHALLENGE);
    cleanup = tmp.cleanup;
    const mod = await importModule<{ removeElements: typeof removeElements }>(
      join(tmp.tmpDir, 'solution.ts'),
    );
    removeElements = mod.removeElements;
  });

  afterAll(() => cleanup());

  it('removes elements from the middle and end', () => {
    expect(toList(removeElements(fromList([1, 2, 6, 3, 4, 5, 6]), 6))).toEqual([1, 2, 3, 4, 5]);
  });

  it('returns null for an empty list', () => {
    expect(removeElements(null, 1)).toBeNull();
  });

  it('removes all elements when all match', () => {
    expect(removeElements(fromList([7, 7, 7, 7]), 7)).toBeNull();
  });

  it('removes elements at the head', () => {
    expect(toList(removeElements(fromList([1, 1, 2, 3]), 1))).toEqual([2, 3]);
  });

  it('returns the same list when no elements match', () => {
    expect(toList(removeElements(fromList([1, 2, 3]), 4))).toEqual([1, 2, 3]);
  });
});
